src/analysis/signature_help.ml

```diff
@@ -65,74 +70,93 @@ let pp_parameter env label ppf ty =
   let param_start = Buffer.length buffer in
   Format.fprintf ppf "%a%!" (pp_parameter env label) ty;
   let param_end = Buffer.length buffer in
   Format.pp_print_string ppf " -> ";
   Format.pp_print_flush ppf ();
-  (param_start, param_end)
+  { label; param_start; param_end; argument }
 
-let application_signature = function
-  (* provide signature information for applied functions *)
-  | (_, Expression earg) :: (_, Expression { exp_desc =
-      Texp_apply ({ exp_type = { desc = Tarrow _; _ }; _ } as efun, eargs); _ }) :: _ ->
-    Type_utils.Printtyp.reset ();
-    let buffer = Buffer.create 16 in
-    let rec separate_signature ?(i=0) ?(param_offsets=[]) ?active_param args ty =
-      match (args, ty) with
-      | [], { Types.desc = Tarrow (label, ty1, ty2, _) } ->
-        let offset = print_parameter_offset buffer efun.exp_env label ty1 in
-        let param_offsets = offset::param_offsets in
-        separate_signature args ty2 ~i:(succ i) ~param_offsets ?active_param
-
-      | arg :: args, { Types.desc = Tarrow (label, ty1, ty2, _) } ->
-        let offset = print_parameter_offset buffer efun.exp_env label ty1 in
-        let param_offsets = offset::param_offsets in
-        let active_param = match arg with
-          | (l, Some e) when l = label && e == earg -> Some i
-          | _ -> active_param
-        in
-        separate_signature args ty2 ~i:(succ i) ~param_offsets ?active_param
-
-      (* end of function type, print remaining type without recording offsets *)
+let separate_function_signature ?args (e : Typedtree.expression) =
+  Type_utils.Printtyp.reset ();
+  let buffer = Buffer.create 16 in
+  let rec separate ?(i=0) ?(parameters=[]) ?args ty =
+    match (args, ty) with
+    | Some ((l, arg) :: args), { Types.desc = Tarrow (label, ty1, ty2, _) } ->
+      let offset = print_parameter_offset buffer e.exp_env label ty1 ?arg in
+      let parameters = offset::parameters in
+      separate ~args ty2 ~i:(succ i) ~parameters
+
+    | _, { Types.desc = Tarrow (label, ty1, ty2, _) } ->
+      let offset = print_parameter_offset buffer e.exp_env label ty1 in
+      let parameters = offset::parameters in
+      separate ty2 ~i:(succ i) ~parameters
+
+    (* end of function type, print remaining type without recording offsets *)
+    | _ ->
+      Format.fprintf (Format.formatter_of_buffer buffer)
+        "%a%!" (pp_type e.exp_env) ty;
+      { fun_name = extract_ident e.exp_desc
+      ; signature = Buffer.contents buffer
+      ; parameters = List.rev parameters
+      ; active_param = None
+      }
+  in
+  separate ?args e.exp_type
+
+let active_parameter_by_arg ~arg params =
+  let find_by_arg = function
+    | { argument = Some a; _ } when a == arg -> true
+    | _ -> false
+  in
+  try Some (List.index params ~f:find_by_arg)
+  with Not_found -> None
+
+let active_parameter_by_prefix ~prefix params =
+  let string_get_opt s i = try Some s.[i] with Invalid_argument _ -> None in
```

Comment:
I'd just check the length here. It's faster and catching `Invalid_argument _` is bad practice.